### PR TITLE
Move to `pyproject.toml` build

### LINKED ISSUE #217: PyPI build failure on github workflow
Wheel build in [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) fails at cmake process.

https://github.com/spglib/spglib/actions/runs/3922329523 (I thought python 3.6 may have problem.)
https://github.com/spglib/spglib/actions/runs/3922368682 (I skipped python 3.6 build, but still fails.)

@LecrisUT, do you have any idea?

### PR DESCRIPTION
**Pull request by LecrisUT**
This PR has a few changes:
- Rename and switch the `ctest` CI to be purely cmake. Didn't make much sense to be python dependent there
- Switch the build to use `pyproject.toml` environment. Some equivalent commands to the `setup.py` ones are:
  - `setup.py build` -> `python -m build` (need to install `pip install build`). This is mostly unnecessary unless you want to fail quickly
  - `setup.py install` -> `pip install .`
- Temporarily removed the `nanoversion` script until we think of how to implement it with dynamic variable scikit-build/scikit-build-core#172

We need to do a few tests on this:
- [x] Test the build CI. Setting the target branch to `develop` initially to trigger these ([CI job finished](https://github.com/spglib/spglib/actions/runs/3924558227))
- [x] Mac build passes with `scikit-build-core`
- [x] Test `pypi-weel-builds`. We'll change the target branch to `rc` if `develop` works to test it there (and if necessary back to `develop` to keep the branches synced)

@atztogo please let me know how the mac build goes. You can manually set it `-DCMAKE_OSX_DEPLOYMENT_TARGET=...` in `pyprojec.toml` if it doesn't work automatically and I'll write a PR upstream for that fix.

About making editable installs, this should be doable via `pip install --editable .`. But for this we need scikit-build/scikit-build-core#171

Closes #217 
Currently, `cibuildwheel` still doesn't work locally. Added upstream issue: scikit-build/scikit-build-core#173. Here's the setup to debug locally assuming you have a docker/podman environment:
```console
$ CIBW_CONTAINER_ENGINE=podman cibuildwheel . --platform linux
```

### CODE REVIEW

**Comment by LecrisUT** [comment]
Worth pointing out. I have tested moving back to the old approach with `setup.py`, but we get again cmake not detecting python in `cibuildwheel`. There is something more fundamentally different here...

Edit: Oh wow, ["the fix"](https://github.com/pypa/cibuildwheel/issues/639#issuecomment-818020937) is so stupid...

**Comment by LecrisUT** [comment]
Sorry about the amount of CI actions. Cmake linking on windows is soo weird

**Comment by LecrisUT** [comment]
@atztogo Finally got it working. Weirdly, the ubuntu tests take half an hour to perform. Any idea on that?

Edit: That was because it was running emulated `aarch64` on `x86_64`

**Comment by atztogo** [comment]
@LecrisUT, thanks for your great effort!

> @atztogo please let me know how the mac build goes. You can manually set it `-DCMAKE_OSX_DEPLOYMENT_TARGET=...` in `pyprojec.toml` if it doesn't work automatically and I'll write a PR upstream for that fix.

I don't see what you mean. What can I do?

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+[build-system]
+requires = ["scikit-build-core", "numpy"]
+build-backend = "scikit_build_core.build"
+
+[project]
+name = "spglib"
+version = "2.1.0-rc1"
```
Comment (lan496) [inline pyproject.toml:7]: Can you add a description at `Contribution.md` that we also need to bump up a version at `pyproject.toml`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.14)
```
Comment (lan496) [inline CMakeLists.txt]: scikit-build-core seems to support CMake>=3.15. So we need 3.15 as well?
https://github.com/scikit-build/scikit-build-core

**Comment by LecrisUT** [comment]
Like what we did in #212. The CIs pass so I suspect it works, but confirm that please

**Review by lan496 — COMMENTED** [review]
@LecrisUT Great works, thanks! I have learned a lot about CMake and Python packaging from your PR. I wrote tiny comments.

File: CMakeLists.txt (hunk 2 above)
Comment (LecrisUT) [reply CMakeLists.txt]: Thanks, didn't notice that

**Comment by LecrisUT** [comment]
Note that last commit is a `[skip ci]` so use merge/squash instead of rebase to trigger the github action again with it uploading to Pypi as well.

**Comment by atztogo** [comment]
> Like what we did in #212. The CIs pass so I suspect it works, but confirm that please

Yes, on macOS arm64, I confirmed that it works. A warning was printed out in the build process by
```
pip install . -vvv
```
which is
```
[ 98%] Building C object python/CMakeFiles/python_api.dir/_spglib.c.o
  /Users/togo/code/spglib/python/_spglib.c:179:1: warning: unknown attribute 'dllexport' ignored [-Wunknown-attributes]
  DLL_PUBLIC PyObject* PyInit__spglib(void)
  ^~~~~~~~~~
  /Users/togo/code/spglib/python/_spglib.c:177:36: note: expanded from macro 'DLL_PUBLIC'
  #define DLL_PUBLIC __attribute__ ((dllexport))
                                     ^~~~~~~~~
  1 warning generated.
```

**Comment by LecrisUT** [comment]
Great, this should be ready now. I've cleaned the warning a bit for more general systems.

Also I've removed the Python2 compatibility macros since it has been sunset-ed for 3 years and we do not test them. Build system would not work for it also.

**Review by lan496 — APPROVED** [review]

**Comment by atztogo** [comment]
Merged to rc. Thanks a lot @LecrisUT!

**Pull request merged**